packages/system/src/system.types.tsx

```diff
@@ -2,15 +2,20 @@ import {
   ResponsiveValue,
   SystemProps,
   SystemStyleObject,
+  ThemeTypings,
 } from "@chakra-ui/styled-system"
 import { Dict } from "@chakra-ui/utils"
 import { Interpolation } from "@emotion/react"
 import * as React from "react"
 
-export interface ThemingProps {
-  variant?: string
-  size?: string
-  colorScheme?: string
+export interface ThemingProps<ThemeComponent extends string = string> {
+  variant?: ThemeComponent extends keyof ThemeTypings["components"]
+    ? ThemeTypings["components"][ThemeComponent]["variants"] | (string & {})
+    : string
+  size?: ThemeComponent extends keyof ThemeTypings["components"]
+    ? ThemeTypings["components"][ThemeComponent]["sizes"] | (string & {})
+    : string
```

Comment:
Yes, I think this would be great as well! This would increase the type safety, but also mean a major bump. Lets test the current approach if it works for the majority of our users and improve it step by step.